### ISSUE TITLE
long / wide output lines "freezing" emacs from ESS

### ISSUE THREAD

**Issue by mmaechler** [emacs-ess/ESS#631]
We've had this issue before I think, but maybe not formally on GH.   I've searched among the open issues and couldn't find it.

Anyway, both the console (I've checked on Linux only) and also RStudio can deal with this easily, but Emacs just basically hangs.. not easily interruptable:

I like to show this with the 'gmp' package (which you should get.. it just "belongs" to a well equipped R).
Then
```r
two <- gmp::as.bigz(2)
fx <- gmp::factorialZ(2 ^ 16)
log10(fx)
```
which will give  `287193.7` telling you that the big integer number has around 287'200 digits.

Printing it in ESS is a real bomb dropped,
but only in ESS.

Using just base R, you get the same effect -- well, this is even larger! -- from
```r
fx <- strrep(paste(LETTERS,letters,collapse=""), 10000)
nchar(fx)
system.time(  print(fx) )
```
where system.time() shows me about 0.1 -- 0.2 seconds when running R in the terminal or also in Rstudio but you have to kill Emacs if you wait too long before killing R inside Emacs, when you do this in ESS ....

Last time this come up, the experts found that this is really a limitation of  Emacs comint,
and that we (ESS developers) cannot do much about it.

Maybe with more modern emacs tools, emacs 26, etc,  some experts among us can try addressing it.
It is just so embarrassing, when other IDE's to R have no problems with this.

If we cannot fix it, I'd still like to keep it as a reminder that this is very very unfortunate.

and possibly  `str(fx)`  before you "bomb" -- in ESS only, not in other GUI:

```
fx
```
will try to display number wi

**Comment by vspinu** [comment]
Indeed, emacs is bad at long lines and there is little we can do about that, but you can `toggle-truncate-lines` in your R buffer (and other files)  and suddenly emacs has no problems with long lines :) I even bound it to C-M-v to quickly switch back and forth. 

As to the actual problem, in my experience it occurs with long strings. I don't know why R prints it in the first place. In any case, I think the real solution for such things is to hide long lines (and long output in general) which would expand gradually when the user clicks on ellipses.

**Comment by lionel-** [comment]
I think we should fix it in base R. Truncating or breaking strings according to `getOption("width")` seems to make sense. The raw deparser should still return the strings in full and without modification but `print()` and auto-print are pretty printers that should take width into account.

**Comment by vspinu** [comment]
I don't think it's ideal; "width" is way too short. Most of the time you would like to inspect your string way further than that. Also I would like to see the end of the string.  

I think the case of data.frames is especially complicated given that columns are right aligned. So it's enough to have one long string and whole data.frame is screwed.

**Comment by lionel-** [comment]
Relevant emacs-devel thread: https://lists.gnu.org/archive/html/emacs-devel/2018-10/msg00470.html

**Comment by mmaechler** [comment]
Thank you for the emacs-devel thread!   I can't get involved there, but I hope they become aware that the issue is not only about editing files with very long lines.  Our case is mostly different (the `*R*` comint process buffer having long lines).   Apropos "whose bug?"  I would strongly declare it an emacs bug, as no other R UI suffers such order of magnitudes of degradation..

**Comment by lionel-** [comment]
I haven't read the whole thing yet, but there seems to be some momentum to provide a minor-mode that disables some features (varying-width characters, character composition, etc) in order to gain speed. There's also a `so-long-mode` that already experiments with this. This may help us for the inferior, but would probably have to be an option because I think we need full unicode support by default.
